### PR TITLE
mbedtls_aes_alt: add setkey for decyption, validate input, fix data type and return type

### PR DESCRIPTION
**Pull request by nkaje**
Let https://github.com/apache/mynewt-core/pull/2175 merge and then merge the top change.

### CODE REVIEW

**Comment by utzig** [comment]
> Let #2175 merge and then merge the top change.

Just merged it!

File: hw/drivers/crypto/src/mbedtls_aes_alt.c [hunk 1]
```diff
@@ -53,15 +54,59 @@ mbedtls_aes_setkey_enc(mbedtls_aes_context *ctx, const unsigned char *key,
...
 int
 mbedtls_aes_crypt_ecb(mbedtls_aes_context *ctx, int mode,
-        const unsigned char input[16], unsigned char output[16])
+                      const unsigned char input[16], unsigned char output[16])
 {
-    return crypto_encrypt_aes_ecb(ctx->crypto, ctx->key, ctx->keylen,
-            (const uint8_t *)input, (uint8_t *)output, AES_BLOCK_LEN);
+    int ret = crypto_encrypt_aes_ecb(ctx->crypto, ctx->key, ctx->keylen,
+                                     (const uint8_t *)input, (uint8_t *)output, AES_BLOCK_LEN);
+    if (ret == AES_BLOCK_LEN) {
```
Comment (utzig) [inline hw/drivers/crypto/src/mbedtls_aes_alt.c:105]: I think `return (ret == AES_BLOCK_LEN) ? 0 : -1;` would look a bit better in this case.

File: hw/drivers/crypto/src/mbedtls_aes_alt.c [hunk 2]
```diff
@@ -53,15 +54,59 @@ mbedtls_aes_setkey_enc(mbedtls_aes_context *ctx, const unsigned char *key,
         return 0;
     }
 
-    return -1;
+    return MBEDTLS_ERR_AES_INVALID_KEY_LENGTH;
+}
+
+int
```
Comment (utzig) [inline hw/drivers/crypto/src/mbedtls_aes_alt.c:60]: Could this be `static`?

**Comment by utzig** [comment]
I am under the impression that `mbedtls_aes_setkey_enc` and `mbedtls_aes_setkey_dec` do the same thing, so maybe the code could be moved into a single function? Nice catch on the error return, the reason for only having `_enc` is that we only use CTR-mode for flash and image encryption and the CTR code in mbedTLS does not check return values!

File: hw/drivers/crypto/src/mbedtls_aes_alt.c [hunk 3]
```diff
@@ -41,8 +42,8 @@ mbedtls_aes_free(mbedtls_aes_context *ctx)
 }
 
 int
```
Comment (utzig) [inline hw/drivers/crypto/src/mbedtls_aes_alt.c:44]: `static` maybe?

File: hw/drivers/crypto/src/mbedtls_aes_alt.c (hunk 1 above)
Comment (nkaje) [reply hw/drivers/crypto/src/mbedtls_aes_alt.c]: done

File: hw/drivers/crypto/src/mbedtls_aes_alt.c (hunk 2 above)
Comment (nkaje) [reply hw/drivers/crypto/src/mbedtls_aes_alt.c]: done

**Comment by nkaje** [comment]
> I am under the impression that `mbedtls_aes_setkey_enc` and `mbedtls_aes_setkey_dec` do the same thing, so maybe the code could be moved into a single function? Nice catch on the error return, the reason for only having `_enc` is that we only use CTR-mode for flash and image encryption and the CTR code in mbedTLS does not check return values!

made a common function.

**Comment by nkaje** [comment]
> LGTM
> 
> The `return (ret == AES_BLOCK_LEN ? 0 : -1);` looks a bit weird to me, because I am very used to `return (ret == AES_BLOCK_LEN) ? 0 : -1;`, but given the low precedence of the ternary conditional it should be fine!

done.

**Pull request merged**